text/0111-auto-resource-detection.md

```diff
@@ -0,0 +1,110 @@
+# Automatic Resource Detection
+
+Describes a mechanism to support auto-detection of resource information.
+
+## Motivation
+
+Resource information, i.e. attributes associated with the entity producing telemetry, can currently be supplied to tracer and meter providers or appended in custom exporters. In addition to this, it would be useful to have a mechanism to automatically detect resource information from the host (e.g. from an environment variable or from aws, gcp, etc metadata) and apply this to traces and metrics. This will in many cases prevent users from having to manually configure resource information.
+
+Note there are some existing implementations of this already in the SDKs (see [below](#prior-art-and-alternatives)), but nothing currently in the specification.
+
+## Explanation
+
+In order to apply auto-detected resource information to traces & metrics, a user will need to:
+
+- Configure which resource detector(s) they would like to run (e.g. AWS EC2 detector)
```

Comment:
Is explicit configuration required? Could we run all cloud providers and take the first one that succeeds? If none succeed then the Detect() call fails.

Reply:
I probably should have added this as an Open question. I've suggested a similar approach to what was in the OpenCensus spec which dicates that cloud-vendor specific detection must not be included in the SDK (and thus would require some configuration), but that may not be necessary. As you point out below, the code to detect instance metadata is usually trivial: lookup an environment variable(s) or get a text response from a known endpoint.

If we do allow cloud vendor-specific detectors to be included in the SDK that raises a few related questions:

- Do we want to enforce that vendor libraries cannot be included in detectors?
- Are we open to any cloud providers adding detection code here as long as they follow the above rule? (other than just major ones)
- Do we attempt detection for all cloud providers by default? (including non-major ones)

Reply:
> Do we want to enforce that vendor libraries cannot be included in detectors?

Yes IMO

> Are we open to any cloud providers adding detection code here as long as they follow the above rule? (other than just major ones)

Yes as long as they add it to OT libraries for all languages.

> Do we attempt detection for all cloud providers by default? (including non-major ones)
If yes, then are there any cases where we would need to let users override this?

Yeah this does seem like it could be tricky given the number of cloud providers and that they're on the rise. Could limit it by some factors, like "must have X% market share" to be included in autodetection but any legitimate provider can still add a provider as a fallback if they don't meet that threshold?

I think ideally we would run all autodetectors in parallel with a relatively small time limit (we use 1s by default in SignalFx agent but I don't have any hard data supporting what the correct value should be).

Reply:
I would start with requiring explicit enumeration of detectors by the user. It can always be extended to an auto-detector.

Reply:
Note the overarching question here is:

"Is it okay for cloud-vendor specific detectors to live in the SDK directly as long as they don't include any third party libraries?"

I'll wait for a couple more opinions before updating the OTEP but if there is general consensus that this is okay, then I'll get rid of the notes about custom detectors living in separate packages.